kafkabp/consumer.go

```diff
@@ -0,0 +1,229 @@
+package kafkabp
+
+import (
+	"context"
+	"io"
+	"sync"
+	"sync/atomic"
+
+	"github.com/Shopify/sarama"
+
+	"github.com/reddit/baseplate.go/log"
+	"github.com/reddit/baseplate.go/metricsbp"
+	"github.com/reddit/baseplate.go/tracing"
+)
+
+// ConsumeMessageFunc is a function type for consuming consumer messages.
+type ConsumeMessageFunc func(ctx context.Context, msg *sarama.ConsumerMessage) error
+
+// ConsumeErrorFunc is a function type for consuming consumer errors.
+type ConsumeErrorFunc func(err error)
+
+// consumer is an instance of a Kafka consumer.
+type consumer struct {
+	cfg ConsumerConfig
+
+	consumer           atomic.Value // sarama.Consumer
+	partitions         atomic.Value // []int32
+	partitionConsumers atomic.Value // []sarama.PartitionConsumer
+
+	closed          int64
+	consumeReturned int64
+
+	wg sync.WaitGroup
+}
+
+// Consumer defines the interface of a consumer struct.
+type Consumer interface {
+	io.Closer
+
+	Consume(ConsumeMessageFunc, ConsumeErrorFunc) error
+
+	// IsHealthy returns false after Consume returns.
+	IsHealthy() bool
+}
+
+// NewConsumer creates a new Kafka consumer.
+func NewConsumer(cfg ConsumerConfig) (Consumer, error) {
+	// Validate input parameters.
+	if cfg.SaramaConfig == nil {
+		cfg.SaramaConfig = DefaultSaramaConfig()
+	}
+	if cfg.ClientID != "" {
+		cfg.SaramaConfig.ClientID = cfg.ClientID
+	}
+	if cfg.SaramaConfig.ClientID == "" {
+		return nil, ErrClientIDEmpty
+	}
+
+	if len(cfg.Brokers) == 0 {
+		return nil, ErrBrokersEmpty
+	}
+
+	if cfg.Topic == "" {
+		return nil, ErrTopicEmpty
+	}
+
+	if cfg.Offset != OffsetNewest {
+		cfg.Offset = OffsetOldest
+	}
+
+	// Return any errors that occurred while consuming on the Errors channel.
+	cfg.SaramaConfig.Consumer.Return.Errors = true
+
+	kc := &consumer{cfg: cfg}
+
+	// Initialize Sarama consumer and set atomic values.
+	if err := kc.reset(); err != nil {
+		return nil, err
+	}
+
+	return kc, nil
+}
+
+func (kc *consumer) getConsumer() sarama.Consumer {
+	c, _ := kc.consumer.Load().(sarama.Consumer)
+	return c
+}
+
+func (kc *consumer) getPartitions() []int32 {
+	p, _ := kc.partitions.Load().([]int32)
+	return p
+}
+
+func (kc *consumer) getPartitionConsumers() []sarama.PartitionConsumer {
+	pc, _ := kc.partitionConsumers.Load().([]sarama.PartitionConsumer)
+	return pc
+}
+
+// reset recreates the consumer and assigns partitions.
+func (kc *consumer) reset() error {
+	if c := kc.getConsumer(); c != nil {
+		if err := c.Close(); err != nil {
+			log.Warnw("Error closing the consumer", "err", err)
+		}
+	}
+
+	rebalance := func() (err error) {
+		c, err := NewSaramaConsumer(kc.cfg.Brokers, kc.cfg.SaramaConfig)
+		if err != nil {
+			return err
+		}
+
+		partitions, err := c.Partitions(kc.cfg.Topic)
+		if err != nil {
+			return err
+		}
+
+		kc.consumer.Store(c)
+		kc.partitions.Store(partitions)
```

Comment:
@fishy This seems kind of weird to me. Using atomic values doesn't seem like the right way to protect these, if reset() got called by 2 separate goroutines at the same time, this would guard concurrent access to each one individually, but could they potentially be out of sync?  Like where goroutine A sets the consumer but goroutine B sets partitions?

This feels like something that should guarded by a `RWMutex` or rather than using `atomic.Value` or is there something I'm missing.

Reply:
so we use individual atomic values mainly for guarding against read and write at the same time. two writes at the same time could happen in theory, but it won't in current implementation (we have only two places calling `reset`, one is in initialization and one is in the infinite loop), so we didn't guard against that.

you can have a struct to store both in the same atomic value. I don't feel strongly either way.